File: crates/aish-scripts/src/executor.rs
use std::collections::HashMap;
use std::process::Command;

use regex::Regex;
use tracing::debug;

use crate::models::Script;

/// Result of executing a .aish script.
#[derive(Debug, Clone)]
pub struct ScriptExecutionResult {
    pub success: bool,
    pub output: String,
    pub error: String,
    pub new_cwd: Option<String>,
    pub env_changes: HashMap<String, String>,
    pub returncode: i32,
}

/// Executes .aish scripts line by line.
pub struct ScriptExecutor {
    /// Optional callback for AI calls within scripts.
    #[allow(clippy::type_complexity)]
    ai_callback: Option<Box<dyn Fn(&str) -> String + Send + Sync>>,
}

impl ScriptExecutor {
    pub fn new() -> Self {
        Self { ai_callback: None }
    }

    pub fn with_ai_callback<F>(self, cb: F) -> Self
    where
        F: Fn(&str) -> String + Send + Sync + 'static,
    {
        Self {
            ai_callback: Some(Box::new(cb)),
        }
    }

    /// Execute a script with the given arguments.
    pub fn execute(&self, script: &Script, args: &[String]) -> ScriptExecutionResult {
        let mut env = self.build_runtime_env(script, args);
        let mut output = String::new();
        let mut env_changes: HashMap<String, String> = HashMap::new();
        let mut current_cwd = env.remove("AISH_CWD").unwrap_or_else(|| "/".to_string());
        let mut returncode = 0;

        let ai_call_re = Regex::new(r#"^\s*ai\s+["']([^"']+)["']\s*$"#).unwrap();
        let return_re = Regex::new(r"^\s*return\s+(.+)$").unwrap();
        let cd_re = Regex::new(r"^\s*cd\s+(.+)$").unwrap();
        let export_re = Regex::new(r"^\s*export\s+([^=]+)=(.*)$").unwrap();
        let ask_re = Regex::new(r#"^\s*ask\s+["']([^"']+)["']\s*$"#).unwrap();

        let lines: Vec<&str> = script.content.lines().collect();
        let mut i = 0;

        // Simple multi-line block detection for if/for/while/case
        let block_start = Regex::new(r"^\s*(if|for|while|case)\b").unwrap();
        let block_end = Regex::new(r"^\s*(fi|done|esac)\b").unwrap();
        let mut in_block = false;
        let mut block_buffer = String::new();

        while i < lines.len() {
            let line = lines[i];
            let trimmed = line.trim();

            i += 1;

            // Skip empty lines and comments
            if trimmed.is_empty() || trimmed.starts_with('#') {
                continue;
            }

            // Handle multi-line blocks
            if in_block {
                block_buffer.push_str(line);
                block_buffer.push('\n');
                if block_end.is_match(trimmed) {
                    in_block = false;
                    // Execute the whole block as bash
                    let result = self.execute_bash(&block_buffer, &current_cwd);
                    output.push_str(&result.output);
                    if !result.success {
                        returncode = result.returncode;
                    }
                    block_buffer.clear();
                }
                continue;
            }

            if block_start.is_match(trimmed) {
                in_block = true;
                block_buffer.push_str(line);
                block_buffer.push('\n');
                continue;
            }

            // Handle return
            if let Some(caps) = return_re.captures(trimmed) {
                if let Some(val) = caps.get(1) {
                    output.push_str(val.as_str());
                }
                break;
            }

            // Handle ai "prompt"
            if let Some(caps) = ai_call_re.captures(trimmed) {
                if let Some(prompt) = caps.get(1) {
                    if let Some(ref cb) = self.ai_callback {
                        let response = cb(prompt.as_str());
                        env.insert("AISH_LAST_OUTPUT".to_string(), response.clone());
                        output.push_str(&response);
                        output.push('\n');
                    } else {
                        output.push_str("(AI not available)\n");
                    }
                }
                continue;
            }

            // Handle ask "prompt"
            if let Some(caps) = ask_re.captures(trimmed) {
                if let Some(question) = caps.get(1) {
                    println!("\x1b[36m{}\x1b[0m", question.as_str());
                    output.push_str("(asked user)\n");
                }
                continue;
            }

            // Handle cd path
            if let Some(caps) = cd_re.captures(trimmed) {
                if let Some(path) = caps.get(1) {
                    let expanded = expand_path(path.as_str().trim(), &current_cwd);
                    if std::path::Path::new(&expanded).is_dir() {
                        current_cwd = expanded;
                    } else {
                        output.push_str(&format!("cd: {}: not a directory\n", expanded));
                    }
                }
                continue;
            }

            // Handle export KEY=VALUE
            if let Some(caps) = export_re.captures(trimmed) {
                if let (Some(key), Some(value)) = (caps.get(1), caps.get(2)) {
                    let key = key.as_str().trim().to_string();
                    let value = value.as_str().trim().to_string();
                    env.insert(key.clone(), value.clone());
                    env_changes.insert(key, value);
                }
                continue;
            }

            // Execute as bash command
            let result = self.execute_bash(trimmed, &current_cwd);
            output.push_str(&result.output);
            if !result.success && returncode == 0 {
                returncode = result.returncode;
            }
        }

        ScriptExecutionResult {
            success: returncode == 0,
            output,
            error: String::new(),
            new_cwd: Some(current_cwd),
            env_changes,
            returncode,
        }
    }

    /// Execute script synchronously as a bash subprocess (for hooks).
    pub fn execute_sync(&self, script: &Script, args: &[String]) -> ScriptExecutionResult {
        let env = self.build_runtime_env(script, args);

        let result = Command::new("/bin/bash")
            .arg("-c")
            .arg(&script.content)
            .envs(&env)
            .output();

        match result {
            Ok(output) => {
                let stdout = String::from_utf8_lossy(&output.stdout).to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).to_string();
                let code = output.status.code().unwrap_or(1);

                // Parse [AISH:CWD:path] markers
                let new_cwd = stdout
                    .lines()
                    .find_map(|line| line.strip_prefix("[AISH:CWD:"))
                    .map(|s| s.trim_end_matches(']').to_string());

                ScriptExecutionResult {
                    success: code == 0,
                    output: stdout,
                    error: stderr,
                    new_cwd,
                    env_changes: HashMap::new(),
                    returncode: code,
                }
            }
            Err(e) => ScriptExecutionResult {
                success: false,
                output: String::new(),
                error: e.to_string(),
                new_cwd: None,
                env_changes: HashMap::new(),
                returncode: 1,
            },
        }
    }

    /// Build runtime environment variables for script execution.
    fn build_runtime_env(&self, script: &Script, args: &[String]) -> HashMap<String, String> {
        let mut env: HashMap<String, String> = std::env::vars().collect();

        env.insert("AISH_SCRIPT_DIR".to_string(), script.base_dir.clone());
        env.insert("AISH_SCRIPT_NAME".to_string(), script.metadata.name.clone());
        if let Ok(cwd) = std::env::current_dir() {
            env.insert("AISH_CWD".to_string(), cwd.to_string_lossy().to_string());
        }

        // Positional args
        for (i, arg) in args.iter().enumerate() {
            env.insert(format!("AISH_ARG_{}", i), arg.clone());
        }

        // Named args from metadata
        for (i, param) in script.metadata.arguments.iter().enumerate() {
            let value = args.get(i).cloned().or_else(|| param.default.clone());
            if let Some(val) = value {
                env.insert(format!("AISH_ARG_{}", param.name.to_uppercase()), val);
            } else if param.required {
                debug!(
                    target: "aish_scripts",
                    "missing required argument: {}", param.name
                );
            }
        }

        env
    }

    /// Execute a bash command and return the result.
    fn execute_bash(&self, command: &str, cwd: &str) -> BashOutput {
        match Command::new("/bin/bash")
            .arg("-c")
            .arg(command)
            .current_dir(cwd)
            .output()
        {
            Ok(output) => {
                let stdout = String::from_utf8_lossy(&output.stdout).to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).to_string();
                let code = output.status.code().unwrap_or(-1);

                let mut result = String::new();
                if !stdout.is_empty() {
                    result.push_str(&stdout);
                }
                if !stderr.is_empty() {
                    if !result.is_empty() {
                        result.push('\n');
                    }
                    result.push_str(&stderr);
                }

                BashOutput {
                    output: result,
                    returncode: code,
                    success: code == 0,
                }
            }
            Err(e) => BashOutput {
                output: format!("Error: {}", e),
                returncode: 1,
                success: false,
            },
        }
    }
}

impl Default for ScriptExecutor {
    fn default() -> Self {
        Self::new()
    }
}

struct BashOutput {
    output: String,
    returncode: i32,
    success: bool,
}

/// Expand ~ and environment variables in a path.
fn expand_path(path: &str, cwd: &str) -> String {
    let expanded = if path.starts_with('~') {
        let home = dirs::home_dir()
            .map(|h| h.to_string_lossy().to_string())
            .unwrap_or_default();
        if let Some(rest) = path.strip_prefix('~') {
            format!("{}{}", home, rest)
        } else {
            path.to_string()
        }
    } else {
        path.to_string()
    };

    // Resolve relative paths
    if std::path::Path::new(&expanded).is_absolute() {
        expanded
    } else {
        format!("{}/{}", cwd.trim_end_matches('/'), expanded)
    }
}

File: src/aish/scripts/executor.py
"""Script executor for running .aish scripts."""

from __future__ import annotations

import logging
import os
import re
import subprocess
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Optional

from ..shell.environment import sanitize_subprocess_loader_env

if TYPE_CHECKING:
    from ..llm import LLMSession
    from ..shell.environment import EnvironmentManager
    from .models import Script

logger = logging.getLogger("aish.scripts.executor")


@dataclass
class ScriptExecutionResult:
    """Result of script execution."""

    success: bool
    output: str = ""
    error: str = ""
    new_cwd: Optional[str] = None
    env_changes: dict[str, str] = field(default_factory=dict)
    returncode: int = 0


class ScriptExecutor:
    """Execute .aish scripts with shell and AI integration."""

    # Pattern to match ai "prompt" calls
    AI_CALL_PATTERN = re.compile(r'^\s*ai\s+"([^"]+)"\s*$')
    # Pattern to match ai 'prompt' calls (single quotes)
    AI_CALL_PATTERN_SINGLE = re.compile(r"^\s*ai\s+'([^']+)'\s*$")
    # Pattern to match return statement
    RETURN_PATTERN = re.compile(r'^\s*return\s+(.+)\s*$')
    # Pattern to match cd command
    CD_PATTERN = re.compile(r'^\s*cd\s+(.+)$')
    # Pattern to match export command
    EXPORT_PATTERN = re.compile(r'^\s*export\s+([^=]+)=(.*)$')
    # Pattern to match ask "prompt" calls
    ASK_PATTERN = re.compile(r'^\s*ask\s+"([^"]+)"\s*$')

    def __init__(
        self,
        llm_session: Optional["LLMSession"] = None,
        env_manager: Optional["EnvironmentManager"] = None,
    ):
        """Initialize the script executor.

        Args:
            llm_session: LLMSession for AI calls.
            env_manager: EnvironmentManager for env variable access.
        """
        self.llm_session = llm_session
        self.env_manager = env_manager

    async def execute(
        self,
        script: "Script",
        args: Optional[list[str]] = None,
        stdin_input: Optional[str] = None,
    ) -> ScriptExecutionResult:
        """Execute a script with given arguments.

        Args:
            script: Script object to execute.
            args: List of string arguments.
            stdin_input: Optional input for interactive ask() calls.

        Returns:
            ScriptExecutionResult with output and state changes.
        """
        args = args or []

        # Build runtime environment
        try:
            runtime_env = self._build_runtime_env(script, args)
        except ValueError as e:
            return ScriptExecutionResult(success=False, error=str(e))

        # Parse and execute script body
        try:
            result = await self._execute_body(
                script.content, runtime_env, stdin_input
            )
            return result
        except Exception as e:
            logger.exception("Script execution failed: %s", e)
            return ScriptExecutionResult(success=False, error=str(e))

    def execute_sync(
        self,
        script: "Script",
        args: Optional[list[str]] = None,
        env: Optional[dict[str, str]] = None,
        timeout: int = 5,
    ) -> ScriptExecutionResult:
        """Execute a script synchronously (for hooks and simple scripts).

        This method runs the script as a simple bash command without
        AI calls or interactive features. Used for hook scripts that
        need to be fast.

        Args:
            script: Script object to execute.
            args: List of string arguments.
            env: Optional extra environment variables to merge.
            timeout: Execution timeout in seconds.

        Returns:
            ScriptExecutionResult with output.
        """
        args = args or []

        # Build runtime environment
        try:
            runtime_env = self._build_runtime_env(script, args)
        except ValueError as e:
            return ScriptExecutionResult(success=False, error=str(e))

        # Merge extra environment variables
        if env:
            runtime_env.update(env)

        # Execute as simple bash script
        try:
            result = subprocess.run(
                script.content,
                shell=True,
                executable="/bin/bash",
                capture_output=True,
                text=True,
                env=runtime_env,
                cwd=runtime_env.get("AISH_CWD", os.getcwd()),
                timeout=timeout,
            )

            output = result.stdout.rstrip("\r\n")
            error = result.stderr.strip() if result.returncode != 0 else ""

            # Parse state changes from output
            new_cwd = None
            env_changes: dict[str, str] = {}

            # Check for cd output pattern
            cd_match = re.search(r"\[AISH:CWD:(.+?)\]", output)
            if cd_match:
                new_cwd = cd_match.group(1)
                output = re.sub(r"\[AISH:CWD:.+?\]", "", output).strip()

            return ScriptExecutionResult(
                success=result.returncode == 0,
                output=output,
                error=error,
                new_cwd=new_cwd,
                env_changes=env_changes,
                returncode=result.returncode,
            )

        except subprocess.TimeoutExpired:
            return ScriptExecutionResult(
                success=False, error=f"Script execution timed out after {timeout}s"
            )
        except Exception as e:
            return ScriptExecutionResult(success=False, error=str(e))

    def _build_runtime_env(
        self, script: "Script", args: list[str]
    ) -> dict[str, str]:
        """Build runtime environment with AISH_ARG_* variables.

        Args:
            script: Script object.
            args: List of string arguments.

        Returns:
            Environment dict with AISH_* variables set.

        Raises:
            ValueError: If required argument is missing.
        """
        env = sanitize_subprocess_loader_env(os.environ)
        env["AISH_SCRIPT_DIR"] = script.base_dir
        env["AISH_CWD"] = os.getcwd()
        env["AISH_SCRIPT_NAME"] = script.name

        # Map arguments to AISH_ARG_<name>
        for i, arg_def in enumerate(script.metadata.arguments):
            arg_name_upper = arg_def.name.upper()
            if i < len(args):
                env[f"AISH_ARG_{arg_name_upper}"] = args[i]
            elif arg_def.default is not None:
                env[f"AISH_ARG_{arg_name_upper}"] = arg_def.default
            elif arg_def.required:
                raise ValueError(
                    f"Missing required argument: {arg_def.name} "
                    f"(use: {script.name} <{arg_def.name}>)"
                )

        # Add positional args as AISH_ARG_0, AISH_ARG_1, etc.
        for i, arg in enumerate(args):
            env[f"AISH_ARG_{i}"] = arg

        # Add exported variables from env_manager
        if self.env_manager:
            env.update(self.env_manager.get_exported_vars())

        return env

    async def _execute_body(
        self,
        content: str,
        env: dict[str, str],
        stdin_input: Optional[str] = None,
    ) -> ScriptExecutionResult:
        """Execute script body line by line.

        Args:
            content: Script content.
            env: Runtime environment.
            stdin_input: Optional stdin for interactive prompts.

        Returns:
            ScriptExecutionResult.
        """
        output_lines: list[str] = []
        new_cwd: Optional[str] = None
        env_changes: dict[str, str] = {}
        current_cwd = env.get("AISH_CWD", os.getcwd())

        lines = content.split("\n")
        i = 0

        while i < len(lines):
            line = lines[i].strip()

            # Skip empty lines and comments
            if not line or line.startswith("#"):
                i += 1
                continue

            # Check for return statement
            if self._match_return(line):
                break

            # Check for ai "prompt" call
            ai_result = self._match_ai_call(line)
            if ai_result:
                try:
                    response = await self._execute_ai_call(ai_result)
                    output_lines.append(response)
                    env["AISH_LAST_OUTPUT"] = response
                except Exception as e:
                    return ScriptExecutionResult(
                        success=False, error=f"AI call failed: {e}"
                    )
                i += 1
                continue

            # Check for ask "prompt" call
            ask_match = self.ASK_PATTERN.match(line)
            if ask_match:
                prompt_text = ask_match.group(1)
                if stdin_input:
                    output_lines.append(stdin_input)
                    env["AISH_LAST_OUTPUT"] = stdin_input
                else:
                    # Simple fallback - in real usage this would prompt user
                    output_lines.append(f"[Interactive prompt: {prompt_text}]")
                i += 1
                continue

            # Check for cd command (capture state change)
            cd_match = self.CD_PATTERN.match(line)
            if cd_match:
                target = cd_match.group(1).strip()
                target = self._expand_path(target, current_cwd, env)

                if os.path.isdir(target):
                    new_cwd = os.path.abspath(target)
                    current_cwd = new_cwd
                    env["AISH_CWD"] = new_cwd
                    env["PWD"] = new_cwd
                    output_lines.append(f"📁 {os.path.basename(new_cwd)}")
                else:
                    return ScriptExecutionResult(
                        success=False, error=f"cd: no such directory: {target}"
                    )
                i += 1
                continue

            # Check for export command
            export_match = self.EXPORT_PATTERN.match(line)
            if export_match:
                key = export_match.group(1).strip()
                value = export_match.group(2).strip().strip("\"'")
                env_changes[key] = value
                env[key] = value
                output_lines.append(f"✅ {key}={value}")
                i += 1
                continue

            # Handle multi-line constructs (if, for, while, etc.)
            if line.endswith(";") or line.startswith(("if ", "for ", "while ", "case ")):
                block_lines = [line]
                i += 1
                # Collect until we see 'fi', 'done', 'esac'
                while i < len(lines):
                    block_line = lines[i]
                    block_lines.append(block_line)
                    if (
                        block_line.strip().startswith("fi")
                        or block_line.strip().startswith("done")
                        or block_line.strip().startswith("esac")
                    ):
                        i += 1
                        break
                    i += 1

                # Execute the block
                result = self._execute_bash("\n".join(block_lines), env, current_cwd)
                if result["output"]:
                    output_lines.append(result["output"])
                if result["error"]:
                    return ScriptExecutionResult(
                        success=False,
                        error=result["error"],
                        output="\n".join(output_lines),
                    )
                continue

            # Execute as bash command
            result = self._execute_bash(line, env, current_cwd)
            if result["output"]:
                output_lines.append(result["output"])
            if result["error"]:
                output_lines.append(f"Error: {result['error']}")

            i += 1

        return ScriptExecutionResult(
            success=True,
            output="\n".join(output_lines),
            new_cwd=new_cwd,
            env_changes=env_changes,
        )

    def _match_return(self, line: str) -> bool:
        """Check if line is a return statement."""
        return bool(self.RETURN_PATTERN.match(line))

    def _match_ai_call(self, line: str) -> Optional[str]:
        """Match ai "prompt" or ai 'prompt' call and return prompt string."""
        match = self.AI_CALL_PATTERN.match(line)
        if match:
            return match.group(1)
        match = self.AI_CALL_PATTERN_SINGLE.match(line)
        if match:
            return match.group(1)
        return None

    async def _execute_ai_call(self, prompt: str) -> str:
        """Execute AI call via LLMSession.

        Args:
            prompt: Prompt string to send to AI.

        Returns:
            AI response text.
        """
        if not self.llm_session:
            return "[AI not available - no LLM session]"

        try:
            response = await self.llm_session.completion(
                prompt=prompt,
                stream=False,
            )
            return response
        except Exception as e:
            logger.error("AI call failed: %s", e)
            raise

    def _execute_bash(
        self, command: str, env: dict[str, str], cwd: str
    ) -> dict[str, Any]:
        """Execute bash command.

        Args:
            command: Bash command string.
            env: Environment dict.
            cwd: Working directory.

        Returns:
            Dict with 'output', 'error', 'returncode'.
        """
        try:
            result = subprocess.run(
                command,
                shell=True,
                executable="/bin/bash",
                capture_output=True,
                text=True,
                env=env,
                cwd=cwd,
            )
            return {
                "output": result.stdout.strip(),
                "error": result.stderr.strip() if result.returncode != 0 else "",
                "returncode": result.returncode,
            }
        except Exception as e:
            return {"output": "", "error": str(e), "returncode": -1}

    def _expand_path(
        self, path: str, current_cwd: str, env: dict[str, str]
    ) -> str:
        """Expand path with ~ and environment variables.

        Args:
            path: Path string to expand.
            current_cwd: Current working directory.
            env: Environment dict.

        Returns:
            Expanded absolute path.
        """
        # Expand ~ to home directory
        path = os.path.expanduser(path)

        # Expand environment variables
        path = os.path.expandvars(path)

        # Make absolute
        if not os.path.isabs(path):
            path = os.path.join(current_cwd, path)

        return os.path.abspath(path)
